contributors/design-proposals/node/cri-windows.md

```diff
@@ -0,0 +1,74 @@
+
+On the Windows platform, processes may be assigned to a job object, which can have [CPU rate control information](https://msdn.microsoft.com/en-us/library/windows/desktop/hh448384(v=vs.85).aspx), memory, and storage resource constraints enforced by the Windows kernel scheduler. A job object is created by Windows to at container creation time so all processes in the container will be aggregated and bound to the resource constraint.
+
+## Umbrella Issue
+[#56734](https://github.com/kubernetes/kubernetes/issues/56734)
+
+## Motivation
+The goal is to start filling the gap of platform support in CRI, specifically for Windows platform. For example, currrently in dockershim Windows containers are scheduled using the default resource constraints and does not respect the resource requests and limits specified in POD. With this proposal, Windows containers will be able to leverage POD spec and CRI to allocate compute resource and respect restriction.
+
+## Proposed design
+
+The design is faily straightforward and to align CRI container configuration for Windows with [OCI runtime specification](https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go):
+```
+// WindowsResources has container runtime resource constraints for containers running on Windows.
+type WindowsResources struct {
+	// Memory restriction configuration.
+	Memory *WindowsMemoryResources `json:"memory,omitempty"`
+	// CPU resource restriction configuration.
+	CPU *WindowsCPUResources `json:"cpu,omitempty"`
+}
+```
+
+Since Storage and Iops for Windows containers is optional, it can be postponed to align with Linux container configuration in CRI. Therefore we propose to add the following to CRI for Windows container (PR [here](https://github.com/kubernetes/kubernetes/pull/57076)).
+
+### API definition
+```
+// WindowsContainerConfig contains platform-specific configuration for
+// Windows-based containers.
+message WindowsContainerConfig {
+    // Resources specification for the container.
+    WindowsContainerResources resources = 1;
+}
+
+// WindowsContainerResources specifies Windows specific configuration for
+// resources.
+message WindowsContainerResources {
+    // CPU shares (relative weight vs. other containers). Default: 0 (not specified).
+    int64 cpu_shares = 1;
+    // Number of CPUs available to the container. Default: 0 (not specified).
+    int64 cpu_count = 2;
+    // Specifies the portion of processor cycles that this container can use as a percentage times 100.
+    int64 cpu_maximum = 3;
+    // Memory limit in bytes. Default: 0 (not specified).
+    int64 memory_limit_in_bytes = 4;
```

Comment:
Could you give an example of how kubernetes cpu/memory requests/limit would map to these fields?

Also, are there any significant difference in these configurations and how kernel reacts to the conditions between linux and windows?

Reply:
> example of how kubernetes cpu/memory requests/limit would map 

```go
CpuCount = int((container.Resources.Limits.Cpu().MilliValue() + 1000)/1000) // 0 if not set
CpuShares = milliCPUToShares(container.Resources.Limits.Cpu().MilliValue())
if CpuShares == 0 {
    CpuShares = milliCPUToShares(container.Resources.Request.Cpu().MilliValue())
}
CpuMaximum =  container.Resources.Limits.Cpu().MilliValue() / sysinfo.NumCPU() / 1000 * 100
if isHyperV {
   CpuMaximum = container.Resources.Limits.Cpu().MilliValue() /CpuCount/1000*100
}
MemoryLimitInBytes = container.Resources.Limits.Memory().Value() 
```

> are there any significant difference in these configurations and how kernel reacts to the conditions between linux and windows?

Those CPU parameters are different from Linux, and there is no CFS on windows. [this MSDN documentation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/resource-controls) explains how cpu/memory resource is controlled for Windows containers.

Reply:
Also to clarify - CpuMaximum is a limits that can't be exceeded. If the node is overprovisioned and under contention for multiple containers operating below their limit, then cycles will be weighted based on shares. If the system isn't overprovisioned then shares has no effect.

Reply:
I see. Thanks for explaining. Those are the things that I wanted to see in this proposal. 

How does CpuCount interact with CpuMaximum?

Reply:
Thanks for reviewing. I'll add to the proposal. 
From https://github.com/docker/docker-ce/blob/master/components/cli/man/docker-run.1.md,
`On Windows Server containers, the processor resource controls are mutually exclusive, the order of precedence is CPUCount first, then CPUShares, and CPUPercent last.`
CPUMaximum is derived from CPUCount for HyperV container, and from sysinfo.NumCPU() for process container.

Reply:
Note that for Hyper-V containers these controls are not mutually exclusive. The CPUCount controls the number of virtual processors that the container has access to (and by extension, the host's logical processors available for execution). The CpuMaximum applies to each of these virtual processors independently, for example, CpuCount=2,CpuMaximum=5000 (50%) would limit each CPU to 50%. Running a single threaded application that uses as much CPU as available on the above configuration would be able to use at most 50% of a single host core.

As mentioned above, for Windows Server containers (process isolation) CpuCount would take precedence (and the container limit would be simulated based on the provided value compared to the number of host CPUs using [JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP](https://msdn.microsoft.com/en-us/library/windows/desktop/hh448384.aspx)), and CpuMaximum would be ignored.

Reply:
@PatrickLang @darrenstahlmsft Thanks for the input. I have added part of your comment into design proposal. Is there any reference to what you have commented besides:
https://github.com/moby/moby/blob/master/daemon/oci_windows.go
https://github.com/Microsoft/hcsshim/blob/master/interface.go
https://github.com/docker/docker-ce/blob/master/components/cli/man/docker-run.1.md
https://msdn.microsoft.com/en-us/library/windows/desktop/hh448384.aspx